File: lib/core/services/session_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';

class SessionService {
  SessionService({
    FirebaseFirestore? firestore,
  })  : _firestore = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _firestore;
  final _uuid = const Uuid();

  String _prefsKey(String uid) => 'active_session_$uid';

  Future<String?> _readLocalSession(String uid) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_prefsKey(uid));
  }

  Future<void> _writeLocalSession(String uid, String sessionId) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_prefsKey(uid), sessionId);
  }

  Future<void> _clearLocalSession(String uid) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_prefsKey(uid));
  }

  Future<String?> _readRemoteSession(String uid) async {
    final doc = await _firestore.collection('users').doc(uid).get();
    final value = doc.data()?['activeSessionId'];
    if (value is! String || value.isEmpty) return null;
    return value;
  }

  Future<void> assertSingleDeviceLogin(String uid) async {
    final remote = await _readRemoteSession(uid);
    if (remote == null) return;

    final local = await _readLocalSession(uid);
    if (local == remote) return;

    throw FirebaseAuthException(
      code: 'session-active',
      message: 'This account is already open on another phone.',
    );
  }

  Future<void> claimSession(String uid) async {
    final sessionId = _uuid.v4();
    await _writeLocalSession(uid, sessionId);
    await _firestore.collection('users').doc(uid).set(
      {
        'activeSessionId': sessionId,
        'activeSessionUpdatedAt': FieldValue.serverTimestamp(),
      },
      SetOptions(merge: true),
    );
  }

  Future<bool> validateLocalSession(String uid) async {
    final remote = await _readRemoteSession(uid);
    if (remote == null) {
      await claimSession(uid);
      return true;
    }

    final local = await _readLocalSession(uid);
    return local != null && local == remote;
  }

  Future<bool> isSessionValid(String uid) async {
    final remote = await _readRemoteSession(uid);
    if (remote == null) return true;
    final local = await _readLocalSession(uid);
    return local != null && local == remote;
  }

  Future<void> clearSession(String uid) async {
    final local = await _readLocalSession(uid);
    final remote = await _readRemoteSession(uid);
    if (local != null && local == remote) {
      await _firestore.collection('users').doc(uid).update({
        'activeSessionId': FieldValue.delete(),
        'activeSessionUpdatedAt': FieldValue.serverTimestamp(),
      });
    }
    await _clearLocalSession(uid);
  }

  Stream<String?> watchRemoteSession(String uid) {
    return _firestore.collection('users').doc(uid).snapshots().map((doc) {
      final value = doc.data()?['activeSessionId'];
      if (value is! String || value.isEmpty) return null;
      return value;
    });
  }
}

File: hilla_ride_ios/HelloTukTuk/Core/Services/SessionService.swift
import FirebaseFirestore
import Foundation

final class SessionService {
    private let firestore = Firestore.firestore()
    private let defaults = UserDefaults.standard

    private func prefsKey(for uid: String) -> String {
        "active_session_\(uid)"
    }

    private func readLocalSession(uid: String) -> String? {
        defaults.string(forKey: prefsKey(for: uid))
    }

    private func writeLocalSession(uid: String, sessionID: String) {
        defaults.set(sessionID, forKey: prefsKey(for: uid))
    }

    private func clearLocalSession(uid: String) {
        defaults.removeObject(forKey: prefsKey(for: uid))
    }

    private func readRemoteSession(uid: String) async throws -> String? {
        let doc = try await firestore.collection("users").document(uid).getDocument()
        guard let value = doc.data()?["activeSessionId"] as? String, !value.isEmpty else {
            return nil
        }
        return value
    }

    func claimSession(uid: String) async throws {
        let sessionID = UUID().uuidString
        writeLocalSession(uid: uid, sessionID: sessionID)
        try await firestore.collection("users").document(uid).setData([
            "activeSessionId": sessionID,
            "activeSessionUpdatedAt": FieldValue.serverTimestamp()
        ], merge: true)
    }

    func clearSession(uid: String) async {
        if let local = readLocalSession(uid: uid),
           let remote = try? await readRemoteSession(uid: uid),
           local == remote {
            try? await firestore.collection("users").document(uid).updateData([
                "activeSessionId": FieldValue.delete(),
                "activeSessionUpdatedAt": FieldValue.serverTimestamp()
            ])
        }
        clearLocalSession(uid: uid)
    }

    func validateLocalSession(uid: String) async throws -> Bool {
        let remote = try await readRemoteSession(uid: uid)
        if remote == nil {
            try await claimSession(uid: uid)
            return true
        }
        guard let local = readLocalSession(uid: uid), local == remote else {
            return false
        }
        return true
    }

    func isSessionValid(uid: String) async -> Bool {
        guard let remote = try? await readRemoteSession(uid: uid) else { return true }
        guard let local = readLocalSession(uid: uid) else { return false }
        return local == remote
    }

    func watchRemoteSession(uid: String) -> AsyncStream<String?> {
        AsyncStream { continuation in
            let listener = firestore.collection("users").document(uid)
                .addSnapshotListener { snapshot, _ in
                    let value = snapshot?.data()?["activeSessionId"] as? String
                    continuation.yield(value?.isEmpty == false ? value : nil)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }
}
